### ISSUE TITLE
TSAN: AnnotateBenignRaceSized lifetime

### ISSUE THREAD

**Issue by olologin** [google/sanitizers#1536]
Hi,
First of all I want to thank you for your efforts in keeping all sanitizers alive, you really help people to fix a lot of software with this.

This "issue" should be qualified as question (I am sorry if this is obvious to you, but I cannot find answer to this).
The question is: Is TSAN automatically "un-annotates" memory region that was previously annotated by AnnotateBenignRaceSized? Or is this [address, address + size] region marked as ignored region till the end of process?

**Comment by dvyukov** [comment]
People should prefer fixing undefined behavior in the code ;)

**Comment by olologin** [comment]
@dvyukov 
Yes, but unfortunately some things are impossible to synchronize without performance regressions.
In our codebase we have one object that gets called from multiple threads to report calculation progress [0, 100]%, which is later reported to user UI. We don't want to implicitly synchronize all threads just because of UI, and correctness of synchronization is not crucial here, performance is.

I am sure you know a lot of such examples in your codebase :)

Anyway, if you think we can remove benign races - I can help with pull request, from what I see it should be simple (But maybe I am wrong).

**Comment by dvyukov** [comment]
Modern C/C++ allow to express that w/o performance loss and w/o making behavior of the program undefined. See std::atomic, _Atomic or __atomic_* builtins.

**Comment by olologin** [comment]
@dvyukov 
You mean storing and reading values from std::atomic with relaxed order?
Yes, that is probably even faster than ignoring races in TSAN.

Although modifying some old legacy code to make it work with std::atomic consumes a lot of time, while annotation is the easiest approach in human-hours sense. I understand that it is UB to not use atomics in this usecase, but I have a feeling that actual behavior of store-load with relaxed order on modern amd64 and arm64 will be equal to behavior of reading/storing normal variables. Maybe I am wrong, if not legacy code I would never use non-atomic variables for such usecase.

But I see your point. I guess we can close this issue if you think there is no need for some inverse operation for AnnotateBenignRaceSized. Thanks for your answers.

**Comment by dvyukov** [comment]
People should prefer fixing undefined behavior but I don't mind merging patch that makes annotations work better provided it does not slow down normal execution and does not complicate code a lot.
But I am not sure how to do it. `Shadow::kRodata` could almost be used for it, but it's not checked on writes and enabling it on writes will slowdown normal execution.
Maybe a "has annotations" flag can shoved into `MBlock` and then it can be checked on free. It should solve it at least for heap objects (but not for stack).

**Comment by olologin** [comment]
>But I am not sure how to do it. 
I was thinking about creating second function in [tsan_interface_ann.cpp](https://github.com/llvm/llvm-project/blob/main/compiler-rt/lib/tsan/rtl/tsan_interface_ann.cpp), let's say "AnnotateBenignRaceSizedEnd", lock dyn_ann_ctx->mtx in it, find our race in benign list, remove it, and exit.
I think this is the easiest approach, which will let us pass test from second post, and will allow users to implement RAII wrapper, plus it will not slowdown/break TSAN.

**Comment by dvyukov** [comment]
Sounds reasonable.
